jmh/jmh.bzl

```diff
@@ -61,6 +61,44 @@ jdk_tool = Label("//tools/defaults:jdk")
 jar_tool = Label("@local_jdk//:jar")
 jar_creator_tool = Label("//src/java/io/bazel/rulesscala/jar:binary")
 
+def _scala_construct_runtime_classpath(deps):
+  scala_targets = [d.scala for d in deps if hasattr(d, "scala")]
+  java_targets = [d.java for d in deps if hasattr(d, "java")]
+  files = []
+  for scala in scala_targets:
+    files += list(scala.transitive_runtime_deps)
+    files += list(scala.transitive_runtime_exports)
+  for java in java_targets:
+    files += list(java.transitive_runtime_deps)
+  return files
+
+def _scala_generate_benchmark(ctx):
+  class_jar = ctx.attr.src.scala.outputs.class_jar
+  classpath = _scala_construct_runtime_classpath([ctx.attr.src])
+  ctx.action(
+      outputs = [ctx.outputs.benchmark_list, ctx.outputs.compiler_hints, ctx.outputs.src_jar],
+      inputs = [class_jar] + classpath,
+      executable = list(ctx.attr._generator.files)[0],
+      arguments = [f.path for f in [class_jar, ctx.outputs.src_jar] + classpath],
+      progress_message = "Generating benchmark code for %s" % ctx.label,
+  )
+  return struct(
+      files=set([ctx.outputs.src_jar])
+  )
+
+scala_generate_benchmark = rule(
+    implementation = _scala_generate_benchmark,
+    attrs = {
+        "src": attr.label(allow_single_file=True, mandatory=True),
+        "_generator": attr.label(executable=True, cfg="host", default=Label("//src/scala/io/bazel/rules_scala/jmh_support:benchmark_generator"))
+    },
+    outputs = {
+        "src_jar": "%{name}.srcjar",
+        "benchmark_list": "resources/META-INF/BenchmarkList",
```

Comment:
I'm worried you will have a name collision if you have two benchmarks in the same file. Can we not put there into the .srcjar or in another jar?

Or at least make sure in a test (perhaps below) where you have two benchmarks in the same BUILD.